pennylane/templates/subroutines/qpe.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the ``QuantumPhaseEstimation`` template.
+"""
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def QuantumPhaseEstimation(unitary, target_wires, estimation_wires):
```

Comment:
It would be good to discuss a bit more whether we want to pass the unitary or its generator, i.e., the Hamiltonian. I'm a bit afraid that we're narrowing the scope to how QPE is used in quantum Monte Carlo. But QPE is used in a lot of places, for example in versions of Shor's algorithm, for preparing approximate ground states of Hamiltonians, and more.

Maybe we can compile a list of applications of QPE and that can help guide how best to create a template around it? :thinking:

Reply:
I really like the suggestion about supporting input Hamiltonians, and yes you're right I hadn't been thinking so much in that direction.

So we have the two options for input:
1. unitary, eventually as a circuit
2. Hamiltonian, input as a `qml.Hamiltonian`, and evolution time `t`, and we can do exp(2 pi i H t).

For me, the best option would be to have two separate templates, given that the inputs and how they will be dealt with are quite different. In option 2, we'd need to use `ApproxTimeEvolution`, whereas option 1 would directly apply controlled versions of the circuit. In either case, we need access to a `qml.control()` transform.

For now, we've done option 1 with a matrix-based input. I'm not sure if it's worth supporting a matrix-based version of option 2, since one could just calculate the matrix exponential themselves. On the other hand, users typically have access to a `qml.Hamiltonian` and I couldn't work out how to get the matrix version of that! But with `qml.Hamiltonian` I can only see how to do it with `qml.control()`.

So overall, I totally agree, but think we'd be better to wait for supporting functionality to come online.

One question might be what we call the two options, e.g.,
1. `QuantumPhaseEstimation` or `UnitaryPhaseEstimation`?
2. `HamiltonianPhaseEstimation`?

Reply:
I'm happy to wait for more functionality to come online before supporting Hamiltonian inputs. I'm not sure if having two templates is preferable to supporting to different types of arguments, but we can discuss that in the future. Re names: let's also discuss in the future. I like keeping `QuantumPhaseEstimation` for this form of the template